### DISCUSSION TITLE
Cross-Database Table Access in SQLAlchemy ORM

### DISCUSSION THREAD

**Discussion by ay27** [sqlalchemy/sqlalchemy#11984] [Usage Questions] [answered]
I encountered a problem using SQLAlchemy's ORM functionality when attempting to access tables across different databases simultaneously. Typically, we define an ORM class like this:

```python
Base = declarative_base()
class A(Base):
    __tablename__ = 'A'
    id = Column(Integer, primary_key=True)
    name = Column(String(32))
```

When we need to operate on two databases at the same time, things get a bit more complex. We can write code like this:

```python
# To avoid __tablename__ conflicts, we need to create two separate registries
mapper_registry1 = registry()
Base1 = mapper_registry1.generate_base()

engine_url_format = "mysql+pymysql://{username}:{password}@{host}:{port}/{database}?charset=utf8"
engine_url1 = engine_url_format.format(database="db1", **db_params)
db_engine1 = create_engine(engine_url1)

# Create the table for the first class
class A(Base1):
    __dbname__ = 'db1'
    __tablename__ = 'A'
    id = Column(Integer, primary_key=True)
    name = Column(String(32))

Base1.metadata.create_all(db_engine1)

mapper_registry2 = registry()
Base2 = mapper_registry2.generate_base()

engine_url2 = engine_url_format.format(database="db2", **db_params)
db_engine2 = create_engine(engine_url2)

# Create the table for the second class
class B(Base2):
    __dbname__ = 'db2'
    __tablename__ = 'A'
    id = Column(Integer, primary_key=True)
    name = Column(String(32))

# Create the B table in the second database
Base2.metadata.create_all(db_engine2)
```

For clarity, I added a `__dbname__` field to differentiate which database they belong to. It is feasible to operate these two ORMs carefully using different engines/sessions.

**My question is, how can we perform a `JOIN` operation between them?**

In contrast, things in GORM is very simple as:

```go
type A1 struct {
    ID   `gorm:"column:id;primaryKey"`
    Name `gorm:"column:name"`
}
func (A1) TableName() string {
    return "db1.A"
}

type A2 struct {
    ID   `gorm:"column:id;primaryKey"`
    Name `gorm:"column:name"`
}
func (A2) TableName() string {
    return "db2.A"
}
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

What you are describing as referred as schema in sqlalchemy. so you would:
- use a single metadata, so a single base instead of two (not required but makes things simpler)
- use the `schema` parameter in the orm model https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#explicit-schema-name-with-declarative-table

**Reply by ay27** [reply]
Thx.
